harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -636,3 +666,71 @@ def test_checkpoint_loading(ckpt: str, istrial: bool):
         assert isinstance(trial, pytorch.PyTorchTrial), type(trial)
     else:
         assert isinstance(trial, torch.nn.Module), type(trial)
+
+
+def make_amp_workloads(
+    trial_class,
+    assert_output_float16=False,
+) -> workload.Stream:
+    trainer = utils.TrainAndValidate()
+    yield from trainer.send(steps=20, validation_freq=1, scheduling_unit=1)
+    training_metrics, _ = trainer.result()
+
+    for metrics, next_metrics in zip(training_metrics[:-1], training_metrics[1:]):
+        scale = next_metrics["scale_before"]
+        if "scale" in metrics:
+            assert metrics["scale"] == scale
+        else:
+            metrics["scale"] = scale
+        assert "scale" in metrics
+    loss_prev = None
+    GROWTH_INTERVAL = trial_class._growth_interval
+    growth_countdown = GROWTH_INTERVAL
+    # Only attempt assertions up to and including the penultimate batch, because
+    #  we may not have the updated scale from the final batch.
+    for idx, metrics in enumerate(training_metrics[:-1]):
+        assert metrics["loss"].dtype is np.dtype("float32")
+        if assert_output_float16:
+            assert metrics["output"].dtype is np.dtype("float16")
+        else:
+            # Automatic usages of Apex or native AMP cast the output back to float32
+            # For the latter case, see the hook end_f16
+            #   defined in PyTorchTrialContext.autocast_forward_pass
+            assert metrics["output"].dtype is np.dtype("float32")
+        loss = metrics["loss"].item()
+        scale_before = metrics["scale_before"]
+        scaled_loss = loss * scale_before
+        if "scaled_loss" in metrics:
+            assert scaled_loss == metrics["scaled_loss"]
+        growth_countdown -= 1
+        scale = metrics["scale"]
+        if scale != scale_before:
+            if scale < scale_before:
+                assert scaled_loss > 32758, (
```

Comment:
can we maybe make this magic number a variable? or at least add a comment so it's less magical?

Reply:
We definitely should, but it's exact meaning is mysterious to me. I found through some trial-and-error that - in the native AMP case (dunno about Apex) - this is the maximum value of the scaled loss for which the scaler is not reduced. Why it's this particular value I don't know. I had expected `65504/2 = 32752` (half the largest normal FP16 number) or `65536/2 = 32768`.

Reply:
I was slightly wrong about the inequality, but I've confirmed that I have it correct now (and that it's the same for Apex), and I also gave a name to the magic number.